### PR TITLE
ch4/ucx: using the ucx nbx am interface

### LINKED ISSUE #6636: jenkins: ch4-ucx-am-only threads/pt2pt/multisend timeout
It happens sporadically. Recently in #6591

### PR DESCRIPTION
**Pull request by hzhou**
## Pull Request Description

The new ucx am interface allows internal rendezvous protocol, which allows the possibility to taking advantage of direct GPU IPC among other benefits. This PR implements ucx am mode using this new am interface.

Fixes #6636
## TODO
* [ ] update the version check to ensure the GPU is supported in the new ucp_am_send_nbx api

## Author Checklist
* [x] **Provide Description** 
      Particularly focus on _why_, not _what_. Reference background, issues, test failures, xfail entries, etc.
* [x] **Commits Follow Good Practice**
      Commits are self-contained and do not do two things at once. 
      Commit message is of the form: `module: short description` 
      Commit message explains what's in the commit.
* [ ] **Passes All Tests**
      Whitespace checker. Warnings test. Additional tests via comments.
* [x] **Contribution Agreement**
      For non-Argonne authors, check [contribution agreement](http://www.mpich.org/documentation/contributor-docs/). 
      If necessary, request an explicit comment from your companies PR approval manager.

### CODE REVIEW

**Comment by hzhou** [comment]
test:mpich/ch4/ucx

**Comment by hzhou** [comment]
test:mpich/ch4/ucx

**Comment by hzhou** [comment]
test:mpich/ch4/ucx

**Comment by hzhou** [comment]
test:mpich/ch4/ucx

**Comment by hzhou** [comment]
test:mpich/ch4/ucx

**Comment by hzhou** [comment]
test:mpich/ch4/ucx

**Comment by hzhou** [comment]
test:mpich/ch4/ucx

**Comment by hzhou** [comment]
test:mpich/ch4/ucx

**Comment by hzhou** [comment]
test:mpich/ch4/most

**Comment by hzhou** [comment]
test:mpich/ch4/most

File: src/mpid/ch4/netmod/ucx/ucx_am.c [hunk 1]
```diff
@@ -66,3 +66,84 @@ void MPIDI_UCX_am_send_callback(void *request, ucs_status_t status)
 
     MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_MPIDI_UCX_AM_SEND_CALLBACK);
 }
+
+#ifdef HAS_UCP_AM_NBX
```
Comment (raffenet) [inline src/mpid/ch4/netmod/ucx/ucx_am.c:70]: nitpick: we tend to use `HAVE_` for these type of macros.

File: src/mpid/ch4/netmod/ucx/ucx_am.h [hunk 2]
```diff
@@ -37,6 +37,59 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_NM_am_isend(int rank,
...
+        data_ptr = (char *) send_buf + header_size;
+
+        MPIR_Memcpy(header, &ucx_hdr, sizeof(ucx_hdr));
+        MPIR_Memcpy((char *) header + sizeof(ucx_hdr), am_hdr, am_hdr_sz);
+
+        MPI_Aint actual_pack_bytes;
+        mpi_errno = MPIR_Typerep_pack(data, count, datatype, 0, data_ptr, data_sz,
+                                      &actual_pack_bytes);
+        MPIR_ERR_CHECK(mpi_errno);
+        MPIR_Assert(actual_pack_bytes == data_sz);
```
Comment (raffenet) [inline src/mpid/ch4/netmod/ucx/ucx_am.h:69]: We could probably use the UCP generic datatype support to skip packing in the netmod and let UCX handle it. Not sure it buys us anything in practice.

File: src/mpid/ch4/netmod/ucx/ucx_am.h [hunk 3]
```diff
@@ -204,16 +258,65 @@ MPL_STATIC_INLINE_PREFIX bool MPIDI_NM_am_check_eager(MPI_Aint am_hdr_sz, MPI_Ai
...
+
+    MPIDIG_get_recv_buffer(&recv_buf, &data_sz, &is_contig, &in_data_sz, rreq);
+    if (!is_contig || in_data_sz > data_sz) {
+        /* non-contig datatype, need receive into pack buffer */
+        /* ucx will error out if buffer size is less than the promised data size,
+         * also use a pack buffer in this case */
+        recv_buf = MPL_malloc(in_data_sz, MPL_MEM_OTHER);
+        MPIR_Assert(recv_buf);
+        MPIDI_UCX_AM_RECV_REQUEST(rreq, pack_buffer) = recv_buf;
+    } else {
```
Comment (raffenet) [inline src/mpid/ch4/netmod/ucx/ucx_am.h:284]: Same comment here about datatype support. I suppose we can work on that separately.

File: src/mpid/ch4/netmod/ucx/ucx_types.h [hunk 4]
```diff
@@ -18,8 +18,9 @@
 #define UCP_PEER_NAME_MAX         HOST_NAME_MAX
 
 /* Active Message Stuff */
-#define MPIDI_UCX_MAX_AM_EAGER_SZ      (16*1024)
+#define MPIDI_UCX_MAX_AM_EAGER_SZ      (8000)   /* internal max RTS 8256 */
```
Comment (raffenet) [inline src/mpid/ch4/netmod/ucx/ucx_types.h:21]: what's the intention of this change?

**Review by raffenet — COMMENTED** [review]
A few comments, but this is pretty good to go.

File: src/mpid/ch4/netmod/ucx/ucx_am.c (hunk 1 above)
Comment (hzhou) [reply src/mpid/ch4/netmod/ucx/ucx_am.c]: Sure. Updated to `HAVE_UCP_AM_NBX`.

File: src/mpid/ch4/netmod/ucx/ucx_am.h (hunk 2 above)
Comment (hzhou) [reply src/mpid/ch4/netmod/ucx/ucx_am.h:69]: Does the UCP generic datatype use IOVs? Which we probably need to check for densities. Yeah, same with the following comment, let's leave to work on it separately.

File: src/mpid/ch4/netmod/ucx/ucx_types.h (hunk 4 above)
Comment (hzhou) [reply src/mpid/ch4/netmod/ucx/ucx_types.h:21]: UCP internal RTS is limited to 8256 bytes -- no idea why the odd number. In our RMA code we decide whether to send in the flattened datatype as extended header based on the `AM_EAGER_SZ`. The extended header will be packed into RTS and blow up the UCP internal limit if the EAGER size is too big.

**Comment by hzhou** [comment]
test:mpich/ch4/ucx

File: src/mpid/ch4/netmod/ucx/ucx_am.h (hunk 2 above)
Comment (raffenet) [reply src/mpid/ch4/netmod/ucx/ucx_am.h:69]: For generic datatypes, it just does packing, but in the UCP library via a set of function pointers we supply.

File: src/mpid/ch4/netmod/ucx/ucx_types.h (hunk 4 above)
Comment (raffenet) [reply src/mpid/ch4/netmod/ucx/ucx_types.h:21]: Is it documented? I see that number as `SEG_SIZE` in the mm and ib transports. My best guess is it is 64 byte header + 8k copy buffer. I don't have a problem with the change. Maybe the comment could be updated to say this is a UCP RTS limit.

**Comment by hzhou** [comment]
test:mpich/ch4/ucx
test:mpich/ch4/gpu/ucx

**Comment by hzhou** [comment]
test:mpich/ch4/ucx

**Comment by hzhou** [comment]
test:mpich/ch4/ucx
test:mpich/ch4/gpu/ucx

**Comment by hzhou** [comment]
test:mpich/ch4/most

**Comment by hzhou** [comment]
test:mpich/ch4/ucx

**Comment by hzhou** [comment]
test:mpich/ch4/ofi

**Comment by hzhou** [comment]
Can't reproduce. So let me test it one more time -
test:mpich/ch4/ofi

**Comment by hzhou** [comment]
`am-only` `truncmsg2` failures tracked in #6641

**Review by raffenet — APPROVED** [review]

**Pull request merged**